archive/imo/imo2021_q1.lean

```diff
@@ -0,0 +1,264 @@
+import tactic data.real.sqrt
+
+noncomputable theory
+
+open real
+
+/-
+# IMO 2021 Q1
+Let n≥100 be an integer. Ivan writes the numbers n, n+1,..., 2n each on different cards.
+He then shuffles these n+1 cards, and divides them into two piles. Prove that at least one
+of the piles contains two cards such that the sum of their numbers is a perfect square.
+# Solution
+We prove the required claim by constructing a triplet of pairwise unequal
+natural numbers a, b, c ∈ [n, 2n], with each sum of each pair being a perfect square. Then, by
+Pigeonhole principle, at least two numbers in the triplet must lie in the same pile, which
+finishes the proof.
+-/
+lemma lower_bound (n : ℕ) (l : ℤ) (hl : 2 * (l : ℝ) ≥ 2 + sqrt (4 + 2 * n)) :
+  (n : ℤ) + 4 * l ≤ 2 * l * l :=
+begin
+  have h₁ : sqrt (4 + 2 * n) ≤ 2 * l - 2 := by linarith,
+  replace h₁ := (sqrt_le_iff.1 h₁).2,
+  norm_cast at h₁, push_cast at h₁,
+  linarith,
+end
+
+lemma upper_bound (n : ℕ) (l : ℤ) (hl₁ : l ≥ 0) (hl₂ : (l : ℝ) ≤ sqrt (1 + n) - 1):
+2 * l * l + 4 * l ≤ 2 * n :=
+begin
+  have h₁ : (l : ℝ) + 1 ≤ sqrt (1 + n) := by linarith,
+  rw le_sqrt at h₁,
+  ring_exp at h₁,
+  norm_cast at h₁, push_cast at h₁,
+  linarith,
+  repeat {norm_cast, linarith},
+end
+
+
+lemma radical_inequality {n : ℕ} (h : n ≥ 107) : sqrt (4 + 2 * n) ≤ 2 * (sqrt (1 + n) - 3) :=
+begin
+  rw sqrt_le_iff,
+  split,
+  { norm_num,
+    rw le_sqrt,
+    repeat {norm_cast, linarith}},
+  ring_exp,
+  rw [pow_two, ← sqrt_mul, sqrt_mul_self],
+  suffices : 24 * sqrt (1 + n) ≤ 2 * n + 36,
+  linarith,
+  rw mul_self_le_mul_self_iff,
+  ring_exp,
+  rw [pow_two, ← sqrt_mul, sqrt_mul_self],
+  --Not splitting into cases lead to a deterministic timeout on my machine
+  by_cases p: n ≥ 108,
+  norm_cast,
+  nlinarith,
+  have : n = 107 := by linarith,
+  subst this,
+  norm_num,
+  repeat {norm_cast, linarith},
+  repeat {norm_num, apply sqrt_nonneg},
+end
+
+/-We will later make use of the fact that there exists (l : ℕ) such that
+n ≤ 2 * l * l - 4 * l and 2 * l * l + 4 * l ≤ 2 * n for n ≥ 107.
+-/
+lemma exists_numbers_in_interval (n : ℕ) (hn : n ≥ 107): ∃ (l : ℤ), l ≥ 0 ∧
+((n : ℤ) + 4 * l ≤ 2 * l * l ∧ 2 * l * l + 4 * l ≤ 2 * n) :=
+begin
+  suffices : ∃ (l : ℤ), l ≥ 0 ∧ 2 * (l : ℝ) ≥ 2 + sqrt (4 + 2 * n) ∧ (l : ℝ) ≤ sqrt (1 + n) - 1,
+  cases this with l t,
+  use l,
+  refine ⟨t.1, ⟨lower_bound n l t.2.1, upper_bound n l t.1 t.2.2⟩⟩,
+  use floor (sqrt (1 + n) - 1),
+  split,
+  { simp only [ge_iff_le],
+    rw floor_nonneg,
+    norm_num,
+    rw le_sqrt,
+    repeat {norm_cast, linarith}},
+  refine ⟨_, floor_le (sqrt (1 + n) - 1)⟩,
+  { suffices : 2 * (sqrt (1 + n) - 2) ≥ 2 + sqrt (4 + 2 * n),
```

Comment:
`suffices` usually gets indented, but that would be pretty awkward here; is it worth rewriting this as a `calc` block?

Reply:
I agree with this, will do so in the next version after the review of the version with everything in ℕ